controller/contour/status_test.go

```diff
@@ -0,0 +1,263 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package contour
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	operatorv1alpha1 "github.com/projectcontour/contour-operator/api/v1alpha1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilclock "k8s.io/apimachinery/pkg/util/clock"
+)
+
+func newCondition(t string, status metav1.ConditionStatus, reason, msg string, lt time.Time) metav1.Condition {
+	return metav1.Condition{
+		Type:               t,
+		Status:             status,
+		Reason:             reason,
+		Message:            msg,
+		LastTransitionTime: metav1.NewTime(lt),
+	}
+}
+
+func TestComputeContourAvailableCondition(t *testing.T) {
+	testCases := []struct {
+		description      string
+		deployConditions []appsv1.DeploymentCondition
+		dsAvailable      int32
+		expect           metav1.Condition
+	}{
+		{
+			description: "deployment progressing, daemonset unavailable",
+			deployConditions: []appsv1.DeploymentCondition{
+				{Type: appsv1.DeploymentProgressing, Status: corev1.ConditionTrue},
+			},
+			dsAvailable: int32(0),
+			expect: metav1.Condition{
+				Type:   operatorv1alpha1.ContourAvailableConditionType,
+				Status: metav1.ConditionUnknown,
+			},
+		},
+		{
+			description: "deployment failure, daemonset unavailable",
+			deployConditions: []appsv1.DeploymentCondition{
+				{Type: appsv1.DeploymentReplicaFailure, Status: corev1.ConditionTrue},
```

Comment:
This is effectively the same as the previous test case, isn't it?

Reply:
The previous is available unknown and this is a different condition type. Is it unnecessary to test for both to ensure the code behaves as expected?

Reply:
It looks a little odd since the logic being tested ignores both conditions, but I don't really mind if you prefer to keep these test cases as they are.

Reply:
@Miciah I'm trying to verify that the fall through case gets hit when the deployment condition is `ReplicaFailure` or `Progressing`:

```
	return metav1.Condition{
		Type:    operatorv1alpha1.ContourAvailableConditionType,
		Status:  metav1.ConditionUnknown,
		Reason:  "ContourUnknown",
		Message: "Contour status unknown.",
	}
```

Let me know if you prefer I explicitly state this in core or if you prefer I remove the test cases.